File: cpp/src/kbrcpp.cpp
// Reverend -- Practical Bayesian Inference with Kernel Embeddings
// Copyright (C) 2013 Lachlan McCalman
// lachlan@mccalman.info

// Reverend is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Reverend is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Reverend.  If not, see <http://www.gnu.org/licenses/>.
#include <iostream>
#include <string>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define EIGEN_DONT_PARALLELIZE
#include "io.hpp"
#include "train.hpp"
#include "preimage.hpp"
#include "cumulative.hpp"
#include "lowrankregressor.hpp"
#include "reducedset.hpp"


void trainReducedSet(const TrainingData& fullData, 
    TrainingData& trainData, Settings& settings)
{
  double frac = settings.data_fraction;
  Eigen::MatrixXd X_b = readNPY(settings.filename_xb);
  Eigen::MatrixXd Y_b = readNPY(settings.filename_yb);
  TestingData testData(fullData.x, fullData.y);
  TrainingData newTrain(fullData.u, fullData.lambda, X_b, Y_b);
  
  
  passthroughTrainSettings<Regressor<RBFKernel>, RBFKernel>(
      newTrain, testData, settings);
  
  std::cout << "reduced set size: " << newTrain.x.rows() << std::endl;
  
  if (settings.scaling_strategy == "optimal")
  {
    findReducedSet<RBFKernel>(newTrain, testData, settings);
  }
  trainData = newTrain;
  //write out reduced set
  writeNPY(trainData.x, settings.filename_xr);
  writeNPY(trainData.y, settings.filename_yr);
}

void trainAlgorithm(const TrainingData& fulldata,
                    TrainingData& data,
                    Settings& settings)
{
  if (settings.data_fraction < 1.0)
  {
    if (settings.scaling_strategy == "random" ||
        settings.scaling_strategy == "optimal")
    {
      trainReducedSet(fulldata, data, settings);
    }
    else
    {
      data = fulldata;
      if (settings.scaling_strategy == "lowrank")
        trainSettings<LowRankRegressor<RBFKernel>, RBFKernel>(data, settings);
      else
        trainSettings<Regressor<RBFKernel>, RBFKernel>(data, settings);
    }
  }
  else
  {
    data = fulldata;
    trainSettings<Regressor<RBFKernel>, RBFKernel>(data, settings);
  }
}


int main(int argc, char** argv)
{
  
  std::cout << "kbrcpp initialised." << std::endl;
  srand(time(NULL));
  auto settings = getSettings(argv[1]);
  
  TrainingData fullData = readTrainingData(settings); 
  TestingData testData = readTestingData(settings);
  TrainingData trainData;
  std::cout << "Training..." << std::endl;
  trainAlgorithm(fullData, trainData, settings);
  //useful sizes for preallocation
  uint n = trainData.x.rows();
  uint m = trainData.u.rows();
  uint s = testData.ys.rows();
  Eigen::MatrixXd weights(s,n);
  
  //Create kernels and algorithm 
  std::cout << "Inferring..." << std::endl;
  Kernel<RBFKernel> kx(trainData.x, settings.sigma_x);
  Kernel<RBFKernel> ky(trainData.y, settings.sigma_y);
  if (settings.inference_type == std::string("filter"))
  {
    Filter<RBFKernel> f(n, m, settings);
    f(trainData, kx, ky, testData.ys, 
      settings.epsilon_min, settings.delta_min, weights);
  }
  else
  {
    if (settings.scaling_strategy == "lowrank")
    {
      LowRankRegressor<RBFKernel> r(n, m, settings);
      r(trainData, kx, ky, testData.ys,
        settings.epsilon_min, settings.delta_min, weights);
    }
    else
    {
      Regressor<RBFKernel> r(n, m, settings);
      r(trainData, kx, ky, testData.ys,
        settings.epsilon_min, settings.delta_min, weights);
    }
  }
  //write out the results 
  writeNPY(weights, settings.filename_weights);

  //Normalise weights
  Eigen::MatrixXd preimageWeights(s,n);
  if (settings.normed_weights)
  {
    preimageWeights = weights;
  }
  else
  {
    std::cout << "Normalizing Weights..." << std::endl;
    computeNormedWeights(trainData.x, weights, kx, settings, preimageWeights);
  }

  //---------Post Processing------//

  //Cumulative Estimates
  bool meanMap = settings.cumulative_mean_map;
  if (settings.cumulative_estimate)
  {
    std::cout << "Estimating Cumulative..." << std::endl;
    Eigen::VectorXd cumulates(testData.ys.rows());
    if (settings.direct_cumulative)
    {
      computeCumulates(trainData,testData, weights, kx, meanMap, cumulates);
    }
    else 
    {
      computeCumulates(trainData,testData, preimageWeights, kx, meanMap, cumulates);
    }
    writeNPY(cumulates, settings.filename_cumulative);
  }
  
  //Quantile Estimates 
  if (settings.quantile_estimate)
  {
    std::cout << "Estimating Quantile..." << std::endl;
    Eigen::VectorXd quantiles(testData.ys.rows());
    double tau = settings.quantile;
    if (settings.direct_cumulative)
    {
      computeQuantiles(trainData,testData,weights,kx,tau,meanMap,quantiles);
    }
    else
    {
      computeQuantiles(trainData,testData,preimageWeights,kx,tau,meanMap,quantiles);
    }
    writeNPY(quantiles, settings.filename_quantile);
  }
  
  //Evaluate the embedding and posterior
  // Eigen::MatrixXd embedding(testData.ys.rows(), testData.xs.rows());
  // Eigen::MatrixXd posterior(testData.ys.rows(), testData.xs.rows());
  Eigen::VectorXd embedding(testData.ys.rows());
  Eigen::VectorXd posterior(testData.ys.rows());
  std::cout << "Evaluating Embedding..." << std::endl;
  computeEmbedding(trainData,testData,weights,kx, embedding);
  std::cout << "Evaluating Log Posterior..." << std::endl;
  computeLogPosterior(trainData,testData,preimageWeights,kx, posterior);
  writeNPY(embedding, settings.filename_embedding);
  writeNPY(posterior, settings.filename_posterior);
  std::cout << "kbrcpp task complete." << std::endl;

  return 0;
}

File: reverend/kbrcpp.py
# Reverend -- Practical Bayesian Inference with Kernel Embeddings
# Copyright (C) 2013 Lachlan McCalman
# lachlan@mccalman.info

# Reverend is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# Reverend is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with Reverend.  If not, see <http://www.gnu.org/licenses/>.

import ConfigParser
import numpy as np
import subprocess
import os.path


class Settings(object):
    def __init__(self, prefix=None):
        self.prefix = prefix
        if prefix is not None:
            self.filename_X = prefix + 'X.npy'
            self.filename_Y = prefix + 'Y.npy'
            self.filename_X_b = prefix + 'X_b.npy'
            self.filename_Y_b = prefix + 'Y_b.npy'
            self.filename_Xr = prefix + 'Xr.npy'
            self.filename_Yr = prefix + 'Yr.npy'
            self.filename_X_s = prefix + 'X_s.npy'
            self.filename_Y_s = prefix + 'Y_s.npy'
            self.filename_U = prefix + 'U.npy'
            self.filename_weights = prefix + 'W.npy'
            self.filename_preimage = prefix + 'PW.npy'
            self.filename_embedding = prefix + 'E.npy'
            self.filename_posterior = prefix + 'P.npy'
            self.filename_cumulative = prefix + 'C.npy'
            self.filename_sgd = prefix + 'SGD.npy'

def write_config_file(settings, filename):
    config = ConfigParser.RawConfigParser()
    config.add_section('Algorithm')
    config.add_section('Input')
    config.set('Input', 'filename_x', settings.filename_X)
    config.set('Input', 'filename_y', settings.filename_Y)
    config.set('Input', 'filename_xr', settings.filename_Xr)
    config.set('Input', 'filename_yr', settings.filename_Yr)
    config.set('Input', 'filename_xs', settings.filename_X_s)
    config.set('Input', 'filename_ys', settings.filename_Y_s)
    config.set('Input', 'filename_xb', settings.filename_X_b)
    config.set('Input', 'filename_yb', settings.filename_Y_b)
    config.set('Input', 'filename_u', settings.filename_U)
    config.add_section('Kernel')
    config.set('Kernel', 'sigma_x', settings.sigma_x)
    config.set('Kernel', 'sigma_y', settings.sigma_y)
    config.set('Kernel', 'sigma_x_min', settings.sigma_x_min)
    config.set('Kernel', 'sigma_y_min', settings.sigma_y_min)
    config.set('Kernel', 'sigma_x_max', settings.sigma_x_max)
    config.set('Kernel', 'sigma_y_max', settings.sigma_y_max)
    config.add_section('Output')
    config.set('Output', 'filename_weights', settings.filename_weights)
    config.set('Output', 'filename_embedding', settings.filename_embedding)
    config.set('Output', 'filename_sgd', settings.filename_sgd)
    config.add_section('Training')
    config.set('Training', 'walltime', settings.walltime)
    config.set('Training', 'folds', settings.folds)
    config.set('Algorithm', 'epsilon_min', settings.epsilon_min)
    config.set('Algorithm', 'delta_min', settings.delta_min)
    config.set('Algorithm', 'epsilon_min_min', settings.epsilon_min_min)
    config.set('Algorithm', 'delta_min_min', settings.delta_min_min)
    config.set('Algorithm', 'epsilon_min_max', settings.epsilon_min_max)
    config.set('Algorithm', 'delta_min_max', settings.delta_min_max)
    config.set('Algorithm', 'normed_weights', int(settings.normed_weights))

    settings.filename_quantile = settings.prefix + 'Q_{}.npy'.format(
        settings.quantile)
    config.set('Algorithm', 'observation_period',
               settings.observation_period)
    config.set('Algorithm', 'inference_type', settings.inference_type)
    config.set('Algorithm', 'cumulative_estimate',
               int(settings.cumulative_estimate))
    config.set('Algorithm', 'cumulative_mean_map',
            int(settings.cumulative_mean_map))
    config.set('Algorithm', 'quantile_estimate',
               int(settings.quantile_estimate))
    config.set('Algorithm', 'pinball_loss',
            int(settings.pinball_loss))
    config.set('Algorithm', 'direct_cumulative',
            int(settings.direct_cumulative))
    config.set('Algorithm', 'quantile', settings.quantile)
    config.set('Output', 'filename_preimage', settings.filename_preimage)
    config.set('Output', 'filename_posterior', settings.filename_posterior)
    config.set('Output', 'filename_cumulative',
               settings.filename_cumulative)
    config.set('Output', 'filename_quantile', settings.filename_quantile)
    config.set('Training', 'preimage_walltime', settings.preimage_walltime)
    config.add_section('Preimage')
    config.set('Preimage', 'preimage_reg', settings.preimage_reg)
    config.set('Preimage', 'preimage_reg_min', settings.preimage_reg_min)
    config.set('Preimage', 'preimage_reg_max', settings.preimage_reg_max)
    config.add_section('Scaling')
    config.set('Scaling', 'scaling_strategy', settings.scaling_strategy)
    config.set('Scaling', 'data_fraction', settings.data_fraction)
    config.set('Scaling', 'sgd_iterations', settings.sgd_iterations)
    config.set('Scaling', 'sgd_learn_rate', settings.sgd_learn_rate)
    config.set('Scaling', 'sgd_batch_size', settings.sgd_batch_size)
    with open(filename, 'w') as configfile:
        config.write(configfile)


def write_data_files(settings,U, X, Y, X_s, Y_s, X_b, Y_b):
    np.save(settings.filename_U, U)
    np.save(settings.filename_X, X)
    np.save(settings.filename_Y, Y)
    np.save(settings.filename_X_s, X_s)
    np.save(settings.filename_Y_s, Y_s)
    np.save(settings.filename_X_b, X_b)
    np.save(settings.filename_Y_b, Y_b)


def run(filename_config, directory):
    path = os.path.join(os.path.abspath(directory), 'kbrcpp')
    proc = subprocess.Popen([path, filename_config],
                            stdout=subprocess.PIPE,
                            stderr=subprocess.STDOUT)
    for line in iter(proc.stdout.readline, ""):
        print line,

